test/construct.test.did

```diff
@@ -0,0 +1,152 @@
+assert blob "DIDL\01\6c\02\00\7c\01\7e\01\00\2a\01" == "(record {42; true})"      : (record {int; bool}) "record: tuple";
+assert blob "DIDL\01\6c\02\00\7c\01\7e\01\00\2a\01" == "(record {1 = true})"      : (record {1:bool}) "record: ignore fields";
+assert blob "DIDL\01\6c\02\00\7c\01\7e\01\00\2a\01"                              !: (record {bool; int}) "record: type mismatch";
+assert blob "DIDL\01\6c\02\00\7c\00\7e\01\00\2a\01"                              !: (record {int; bool}) "record: duplicate fields";
+assert blob "DIDL\01\6c\02\01\7c\00\7e\01\00\2a\01"                              !: (record {1:int; 0:bool}) "record: unsorted";
+assert blob "DIDL\01\6c\02\d3\e3\aa\02\7e\86\8e\b7\02\7c\01\00\01\2a"
+    == "(record { foo = 42; bar = true})"                                         : (record {foo:int; bar:bool}) "record: named fields";
+assert blob "DIDL\01\6c\01\cd\84\b0\05\7f\01\00" == "(record { \"☃\" = null })"   : (record { "☃":null }) "record: unicode field";
+assert blob "DIDL\01\6c\01\80\e4\97\d0\12\7c\01\00\2a"                         !: (record {}) "record: field hash larger than u32";
+assert blob "DIDL\04\6c\01\00\01\6c\01\00\02\6c\01\00\03\6c\00\01\00"
+    == "(record { record { record { record {} } } })"                           : (record{record{record{record{}}}}) "record: nested";
+assert blob "DIDL\01\6c\01\00\00\01\00"                                        !: (EmptyRecord) "record: empty recursion";
+assert blob "DIDL\01\6c\02\00\7c\01\7e\01\00\2a"                               !: (record {int}) "record: value too short";
+assert blob "DIDL\01\6c\02\00\7c\01\7e\01\00\2a\01\00"                         !: (record {int}) "record: value too long";
+assert blob "DIDL\01\6c\02\00\7c\01\01\00\2a\01"                               !: (record {}) "record: type too short";
+assert blob "DIDL\01\6c\02\00\7c\01\00\2a\01"                                  !: (record {}) "record: type too short";
+assert blob "DIDL\01\6c\02\00\7c\01\7e\02\7e\01\00\2a\01\00"                   !: (record {}) "record: type too long";
+assert blob "DIDL\01\6c\02\00\01\01\7e\01\00\00\00"                            !: (record {}) "record: type out of range";
+assert blob "DIDL\01\6c\02\00\01\01\7e\01\7c\2a"                               !: (reserved) "record: type out of range";
+
+// variant
+assert "(variant {})"                                                          !: (variant {}) "variant: no empty value";
+assert blob "DIDL\01\6b\00\01\00"                                              !: (variant {}) "variant: no empty value";
+assert blob "DIDL\01\6b\01\00\7f\01\00\00" == "(variant {0})"                   : (variant {0}) "variant: numbered field";
+assert blob "DIDL\01\6b\01\00\7f\01\00\00\2a"                                  !: (variant {0:int}) "variant: type mismatch";
+
+// TODO Should this pass?
+assert blob "DIDL\01\6b\02\00\7f\01\7c\01\00\01\2a"                             : (variant {0:int; 1:int}) "??? variant: type mismatch";
```

Comment:
Sorry, too lazy to decode. What is unclear here?

Reply:
The blob says `variant { 1 = 42 } : variant { 0: null; 1: int }`, and we want to decode at type `variant { 0:int; 1:int}`. Do we want to care variant type mismatch when the value doesn't use that branch?

Reply:
Oh, right. That was (is?) discussed in https://github.com/dfinity-lab/motoko/issues/870. I also very much vote against requiring every decode to implement a full subtyping checker, which is a non-trivial piece of code, especially for recursive types.

Reply:
Let’s continue in https://github.com/dfinity-lab/motoko/pull/1830 maybe